cpp/src/arrow/util/async_nursery.cc

```diff
@@ -0,0 +1,130 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/util/async_nursery.h"
+
+#include "arrow/util/logging.h"
+
+namespace arrow {
+namespace util {
+
+AsyncCloseable::AsyncCloseable() = default;
+AsyncCloseable::AsyncCloseable(AsyncCloseable* parent) {
+  parent->AddDependentTask(OnClosed());
+}
+
+AsyncCloseable::~AsyncCloseable() {
+  // FIXME - Would be awesome if there were a way to enforce this at compile time
+  DCHECK_NE(nursery_, nullptr) << "An AsyncCloseable must be created with a nursery "
+                                  "using MakeSharedCloseable or MakeUniqueCloseable";
+}
+
+const Future<>& AsyncCloseable::OnClosed() {
+  // Lazily create the future to save effort if we don't need it
+  if (!on_closed_.is_valid()) {
+    on_closed_ = Future<>::Make();
+  }
```

Comment:
I'm a bit worried about this being race-prone. So far it looks like it's only used in the one constructor, perhaps it could be inlined there?